### ISSUE TITLE
node.js deployment: Error MSB3428: Could not load the Visual C++ component "VCBuild.exe"

### ISSUE THREAD

**Issue by bpasero** [projectkudu/kudu#1076]
Issue:
npm deployment produces the following error

```
remote: Building the projects in this solution one at a time. To enable parallel build, please add the "/m" switch.
remote: MSBUILD : error MSB3428: Could not load the Visual C++ component "VCBuild.exe". To fix this, 1) install the .NET Framework 2.0 SDK, 2) install Microsoft Visual Studio 2005 or 3) add the location of the component to the system path if it is installed elsewhere.  [D:\home\site\wwwroot\node_modules\tsd\node_modules\bufferstream\node_modules\buffertools\build\binding.sln]
```

To reproduce:
- enable git for a azure web site
- clone locally
- add a package.json (see contents below)
- git push to trigger deployment

package.json
```
{
	"name": "s",
	"description": "s",
	"version": "0.6.0",
	"author": {
    	"name": "Microsoft Corporation"
 	},
	"private": true,
	"engines": { "node": ">= 0.10.0 < 0.11.0" },
	
	"dependencies": {
		"tsd": "*"
	}
}
```

**Comment by xdansmith** [comment]
Are there any plans to remove this limitation? MSBuild is present, so I don't see why the VC++ build tools couldn't be as well.

We're trying to move over to Azure Websites but it seems running node.js is a little trickier than we originally thought because of limitations like this.

**Comment by phairoh** [comment]
Every time that I want to love Azure I run in to issues like this and end up struggling for hours trying to figure out work arounds.

This seems like a very silly limitation as all it would take to fix it is installing the appropriate build tools.  Is there any plan to fix this or are we left to figure out workarounds indefinitely or switch away from Azure to a competing cloud computing provider?

**Comment by adambuczynski** [comment]
+1

**Comment by DavidRouyer** [comment]
This is a blocker issue for a lot of people. It is almost impossible to build a decent workflow or a decent application without one or more native modules.

Nobody wants to put hundred on megabytes of node_modules dependencies inside their source control repository.
And how can we build a continuous deployment script if we can't put the "npm install" command inside our scripts ? (obviously npm install is failing due to node_gyp modules)

Please add VC++ build tools ! We don't want workarounds, we want a first class user experience when using Node.js on Azure !

**Comment by francesconero** [comment]
Yes, being able to build native modules would remove so many workarounds needed to deploy a complex node.js project!

**Comment by DavidRouyer** [comment]
Can we have an update on that ?

**Comment by davidebbo** [comment]
Sorry, there is no useful update at this time. We know it's a pain point, and we're still hoping there will be a solution in the future.

**Comment by DavidRouyer** [comment]
Is it not possible to install a VC++ compiler on the platform ?

**Comment by davidebbo** [comment]
There is more to it. It's not just a matter of installing one thing and having everything start to work. We did experiment with this at the time and there were several additional issues.

**Comment by davidebbo** [comment]
@francesconero @DavidRouyer Could you please share the name of specific native node modules that you are hitting this issue with?

Some that we already know about:

- socket.io
- bufferstream

**Comment by francesconero** [comment]
The ones I have problems with both come from the mongodb package (although one recently became an optional dependency):

- bson-ext
- kerberos

**Comment by DavidRouyer** [comment]
Actually, every module depending on [node-gyp](https://github.com/nodejs/node-gyp) in my opinion. Your can find a list [here](https://www.npmjs.com/browse/depended/node-gyp).

Take dependencies of dependencies into account.
For example, karma and protractor need socket.io.

Database drivers (sqlite3, pg/ng-native, node-oracle, node-mssql (not sure), etc.) ...

**Comment by davidebbo** [comment]
@DavidRouyer we're looking for *specific* ones that we can use for testing, and we'd want to prioritize the testing with modules that we know people are actually trying to use, so not the full list of everything that uses node-gyp. We'll look at the specific ones you mentioned, thanks!

**Comment by davidebbo** [comment]
Other question: if we had a solution that only worked with npm 3.x, would that be acceptable? Or are people still using the old one?

**Comment by francesconero** [comment]
I think it would be awesome! I'm sure there are still people using 2.x versions (me, for example), and the main problem I see is that peerDependencies were removed in npm 3.x.

But I mean, what better occasion to migrate the install than this? Also I think most projects would upgrade from npm 2.x to 3.x without even touching the package.json.

May I ask a question out of curiosity though? Does the npm3 requirement stem from the path length limitation?

**Comment by DavidRouyer** [comment]
I think npm 2.11.x or later is all good !
I think a lot of people have made the switch to node v4.
For people who wants stability, they can use the v4 LTS branch shipped with npm 2.14.x by default.
For people who wants to use the latest version of node, they can use the v4 branch with npm 2.14.x or npm 3. (The v5 branch is coming very soon and will be shipped with npm 3 by default).

**Comment by ahmelsayed** [comment]
The requirement comes from the [supported versions of VS in node-gyp](https://github.com/nodejs/node-gyp/blob/master/gyp/pylib/gyp/MSVSVersion.py#L228-L320) that shipped with npm in a given version. VS 2015 started being supported by node-gyp that shipped in npm 2.11.2 while VS 2013 support shipped with npm 1.3.21. Ideally we will support one version of the compiler, either 2013 or 2015, but we wanted to see if it's okay to only support newish npm releases (>= 2.11.2)

**Comment by francesconero** [comment]
@ahmelsayed makes sense, thanks! But why jump to npm 3.x, which introduces unrelated breaking changes?

**Comment by ahmelsayed** [comment]
We just thought that 2015 was only supported in 3.x, but apparently it's supported starting from 2.11.2, David had it initially saying 3.x but has updated his comment since then to say 2.11 :)

**Comment by francesconero** [comment]
Of course! Sorry to have missed that. I wish they sent notifications for updates to comments you replied to :)

**Comment by davidebbo** [comment]
One more thing we could use from you: if you could share some simple repos that use those packages and are currently blocked by it, we would use that to validate that things work end to end. Ideally, it should be a simple repo such that:

- pushing it today results in the VCBuild.exe issue
- it should work fully if native modules can be successfully compiled

**Comment by francesconero** [comment]
@davidebbo should the project be a website? The native modules I use are used to interat with mongodb, so you would have to set up an instance before the project worked. It wouldn't be just a simple deploy unfortunately.

**Comment by davidebbo** [comment]
Another approach: email me (david.ebbo (AT) microsoft.com), and we can get you set up on a test site that has the fix, where you can try whatever you need.

**Comment by davidebbo** [comment]
The fix we made is deployed, and should allow npm native compilation to work. Some notes:

- There is a known perf issue that causes the compiler server process (`mspdbsrv.exe`) to hang for about a minute on first invocation.
- We've tested the following npm packages: https://github.com/KuduApps/NodeAppWithNativeNpmModules/blob/master/package.json.
- For a reason that is not fully understood yet, a subset of those packages (hiredis, bcrypt, libxmljs, fibers, node-expat, kerberos) fail to install on free/shared Web Apps, but install fine on dedicated Web Apps (i.e. basic/standard).
- Let us know if you run into an issue with a specific package.

**Comment by Benjiiim** [comment]
Hi @davidebbo,

We're having an issue with the following module which is not in the list of tested packages:  [lwip](https://github.com/EyalAr/lwip).
Repro and deployment logs can be found here: https://github.com/Benjiiim/azure-appservice-test

Any advice if this should work or if we need to install more dependencies through the custom script, would be greatly appreciated.

Thanks

**Comment by davidebbo** [comment]
@Benjiiim can you open a new issue? It will make it easier to track this one case instead of this long thread.

**Comment by nestenborg** [comment]
[lwip_logs.txt](https://github.com/projectkudu/kudu/files/596871/lwip_logs.txt)

**Comment by nestenborg** [comment]
attached log file from the build in the web app, I also have the npm-debug.log if that is required.

thanks
